### DISCUSSION TITLE
a lot of sources, transforms

### DISCUSSION THREAD

**Discussion by Rentu** [vectordotdev/vector#22084] [Q&A]
I have a lot of log collection, and the simplest way to manage them is to use the filename as a source. However, this may lead to too many sources and impact performance, so we do some merging. But in the transform stage, different logs may have different conditions for multi-line matching or different rules for timestamp extraction. This could lead to a large number of transforms, possibly over a thousand. What are some good solutions for such a scenario? Our daily log collection volume is more than 30TB, and the logging scenarios are complex, with inconsistent encoding and a large number of log collections, making the cleansing rules complex.

**Comment by pront** [comment]
> But in the transform stage, different logs may have different conditions for multi-line matching or different rules for timestamp extraction. 

Hmm, this is a challenging task and I am sure I am lacking context based on the description. I am thinking if standard consolidation/grouping practices help here, e.g. normalize logs early in the process and introduce some patterns to check against.

If you add more details and examples maybe we can come up with better recommendations.

**Reply by Rentu** [reply]
Is there a limit on the number of sources, transforms, and pipelines? Or are there any performance testing data or recommendations, such as keeping below a certain quantity level to achieve good performance?

**Reply by jszwedko** [reply]
There is no limit on the number of components. I think it is largely a matter of making sure that Vector has sufficient resources to process whatever config. If you are running Vector with, say, 1000 components and 1 CPU you are likely to see the CPU be a bottleneck. You can find what recommendations we do have for sizing, [here](https://vector.dev/docs/setup/going-to-prod/sizing/), though I'm not sure they address this specific question. I'm unfortunately not sure that we have general recommendations for number of components for a given number of resources. Many people end up doing their own testing / benchmarking.

**Reply by pront** [reply]
> Many people end up doing their own testing / benchmarking.

My two cents on the above; many people end up doing their own testing/benchmarking, and this is desirable because our benchmarks will serve as a guide rather than an authoritative source. Users have full context about their configurations and setups and understand their deployments (hardware, OS, environment) better. This should provide them with more accurate numbers.
